go/runtime/task.go

```diff
@@ -111,6 +108,19 @@ func newTaskBase[TaskSpec pf.Task](
 	}, nil
 }
 
+// StartTaskHeartbeatLoop begins a long-lived task which writes stats at regular intervals,
+// and then logs the final exit status of this shard. The `usageRate` is the number of
+// "credits" used per second by the connector.
+func (t *taskBase[TaskSpec]) StartTaskHeartbeatLoop(shard consumer.Shard, container *pr.Container) {
+	var usageRate float32 = 0.0
+	if container != nil {
+		usageRate = container.UsageRate
+	} else {
+		logrus.WithField("shard", shard.Spec().Id).Info("usageRate will be 0 because there is no container")
```

Comment:
When can `container` be `nil`?

Reply:
Ah, that's what I was trying to get at with [this comment](https://github.com/estuary/flow/pull/1403#issuecomment-1981746981), which wasn't very clear. If the connector type is anything but `Image`, then `container` will be nil. We may need to revisit this if we ever want to charge for other connector types, but for now those are just defaulted to a 0 `usageRate`.